File: frontend/src/lib/api/member.ts
import { API_URL } from "$lib/config";
import { CrudRepo } from "$lib/crudRepo.svelte";
import { okorerr } from "$lib/util";

export interface MemberNew {
  firstname: string;
  lastname: string;
  email: string;
  plz: string;
  city: string;
  street: string;
  housenumber: string;
  iban: string;
  bic: string;
  sepa: boolean;
  note: string | undefined;
  joindate: number;
  birthday: number;
  mandate: string;
  fee: number;
}

export interface Member extends MemberNew {
  id: number;
  paid: boolean;
  deleted: boolean;
}

const MEMBER_URL = API_URL + "member";

export async function getMembers(): Promise<Member[]> {
  return fetch(MEMBER_URL, { credentials: "include" })
    .then(okorerr)
    .then((res) => res.json())
    .then((res: Member[]) => {
      res.forEach((m) => {
        m.joindate *= 1000;
        m.birthday *= 1000;
      });
      return res;
    });
}

export async function getDeletedMembers(): Promise<Member[]> {
  return fetch(MEMBER_URL + "?show_deleted=true", { credentials: "include" })
    .then(okorerr)
    .then((res) => res.json())
    .then((res: Member[]) => {
      res.forEach((m) => {
        m.joindate *= 1000;
        m.birthday *= 1000;
      });
      return res;
    });
}

export async function getMember(id: number): Promise<Member> {
  return fetch(MEMBER_URL + `?id=${id}`, { credentials: "include" })
    .then(okorerr)
    .then((res) => res.json())
    .then((res) => res[0])
    .then((res: Member) => {
      res.joindate *= 1000;
      res.birthday *= 1000;
      return res;
    });
}

export async function postMember(member: MemberNew): Promise<Response> {
  member.joindate /= 1000;
  member.birthday /= 1000;
  return fetch(MEMBER_URL, {
    method: "POST",
    body: JSON.stringify(member),
    headers: { "Content-Type": "application/JSON" },
    credentials: "include",
  }).then(okorerr);
}

export async function deleteMember(id: number): Promise<Response> {
  return fetch(MEMBER_URL + `?id=${id}`, {
    method: "DELETE",
    credentials: "include",
  }).then(okorerr);
}

export async function putMember(id: number, member: MemberNew) {
  member.joindate /= 1000;
  member.birthday /= 1000;
  return fetch(MEMBER_URL + `?id=${id}`, {
    method: "PUT",
    body: JSON.stringify(member),
    headers: { "Content-Type": "application/JSON" },
    credentials: "include",
  }).then(okorerr);
}

export async function restoreMember(id: number) {
  return fetch(MEMBER_URL + `/restore?id=${id}`, {
    credentials: "include",
    method: "POST",
  });
}

export function makeMemberRepo() {
  return new CrudRepo({
    get: getMembers,
    select: getMember,
    add: postMember,
    update: putMember,
    delete: deleteMember,
  });
}

// calling delete on this repo will restore the member / remove the deleted flag
export function makeDeletedMemberRepo() {
  return new CrudRepo({
    get: getDeletedMembers,
    delete: restoreMember,
  });
}

File: src/member.rs
use axum::{extract::Query, http::StatusCode, Json};
use chrono::Datelike;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use crate::AppState;

#[derive(Deserialize, Serialize, FromRow)]
pub struct Member {
    id: i32,
    firstname: String,
    lastname: String,
    email: String,
    plz: String,
    city: String,
    street: String,
    housenumber: String,
    iban: String,
    bic: String,
    sepa: bool,
    note: Option<String>,
    joindate: i32,
    paid: bool,
    birthday: i32,
    fee: f32,
    mandate: String,
    deleted: bool,
}

#[derive(Deserialize, Serialize)]
pub struct MemberNew {
    firstname: String,
    lastname: String,
    email: String,
    plz: String,
    city: String,
    street: String,
    housenumber: String,
    iban: String,
    bic: String,
    sepa: bool,
    note: Option<String>,
    joindate: i32,
    birthday: i32,
    fee: f32,
    mandate: String,
}

#[derive(Deserialize, Serialize)]
pub struct MemberIdQuery {
    id: i32,
}

#[derive(Deserialize, Serialize)]
pub struct ShowDeletedQuery {
    show_deleted: bool,
}

pub async fn get_member(
    state: AppState,
    id: Option<Query<MemberIdQuery>>,
    show_deleted: Option<Query<ShowDeletedQuery>>,
) -> Result<Json<Vec<Member>>, StatusCode> {
    let current_year = chrono::Utc::now().year();

    let where_clause = if let Some(ref id_query) = id {
        format!("where m.id = {}", id_query.id)
    } else {
        format!(
            "where m.deleted = {}",
            show_deleted.map_or("0", |f| if f.show_deleted { "1" } else { "0" })
        )
    };

    let res = sqlx::query_as::<_, Member>(&format!(
        "select
        m.*,
        {current_year} in (select p.year from payment as p where memberid = m.id) as paid
        from member as m
        {where_clause}
        order by firstname asc"
    ))
    .fetch_all(&state.db)
    .await
    .unwrap();

    if id.is_some() && res.is_empty() {
        return Err(StatusCode::NOT_FOUND);
    }

    Ok(Json(res))
}

pub async fn post_member(state: AppState, n: Json<MemberNew>) -> StatusCode {
    sqlx::query(
        "
                   insert into member 
                   (firstname,
                    lastname,
                    email,
                    plz,
                    city,
                    street,
                    housenumber,
                    iban,
                    bic,
                    sepa,
                    note,
                    joindate,
                    birthday,
                    fee,
                    mandate
                   )
                   values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
    )
    .bind(&n.firstname)
    .bind(&n.lastname)
    .bind(&n.email)
    .bind(&n.plz)
    .bind(&n.city)
    .bind(&n.street)
    .bind(&n.housenumber)
    .bind(&n.iban)
    .bind(&n.bic)
    .bind(n.sepa)
    .bind(&n.note)
    .bind(n.joindate)
    .bind(n.birthday)
    .bind(n.fee)
    .bind(&n.mandate)
    .execute(&state.db)
    .await
    .unwrap();

    StatusCode::OK
}

pub async fn delete_member(state: AppState, q: Query<MemberIdQuery>) {
    sqlx::query("update member set deleted=1 where id = ?")
        .bind(q.id)
        .execute(&state.db)
        .await
        .unwrap();
}

pub async fn put_member(state: AppState, q: Query<MemberIdQuery>, n: Json<MemberNew>) {
    sqlx::query(
        "update member set
                   firstname=?,
                    lastname=?,
                    email=?,
                    plz=?,
                    city=?,
                    street=?,
                    housenumber=?,
                    iban=?,
                    bic=?,
                    sepa=?,
                    note=?,
                    joindate=?,
                    note=?,
                    birthday=?,
                    fee=?,
                    mandate=?
        where id=?
                ",
    )
    .bind(&n.firstname)
    .bind(&n.lastname)
    .bind(&n.email)
    .bind(&n.plz)
    .bind(&n.city)
    .bind(&n.street)
    .bind(&n.housenumber)
    .bind(&n.iban)
    .bind(&n.bic)
    .bind(n.sepa)
    .bind(&n.note)
    .bind(n.joindate)
    .bind(&n.note)
    .bind(n.birthday)
    .bind(n.fee)
    .bind(&n.mandate)
    .bind(q.id)
    .execute(&state.db)
    .await
    .unwrap();
}

pub async fn restore_member(state: AppState, q: Query<MemberIdQuery>) {
    sqlx::query("update member set deleted=0 where id=?")
        .bind(q.id)
        .execute(&state.db)
        .await
        .unwrap();
}
